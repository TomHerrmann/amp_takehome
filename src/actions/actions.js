import {
  API_ERROR,
  CONTACTS_POPULATE,
  CONTACT_SELECT,
} from '../constants/actionTypes';

export const apiError = (err) => ({
  type: API_ERROR,
  payload: err,
});

export const contactsPopulate = (contacts) => ({
  type: CONTACTS_POPULATE,
  payload: contacts,
});

export const contactSelect = (contact) => ({
  type: CONTACT_SELECT,
  payload: contact,
});
