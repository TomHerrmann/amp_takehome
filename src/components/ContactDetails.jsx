import React, { useCallback, useState } from 'react';
import { format } from 'path';

const ContactDetails = ({ contact }) => {
  const { firstName, lastName, address, city, country, phone, email } = contact;

  const [formFirstName, setFormFirstName] = useState(firstName);
  const [formLastName, setFormLastName] = useState(lastName);
  const [formAddress, setFormAddress] = useState(address);
  const [formCity, setFormCity] = useState(city);
  const [formCountry, setFormCountry] = useState(country);
  const [formPhone, setFormPhone] = useState(phone);
  const [formEmail, setFormEmail] = useState(email);

  const formFields = [
    'First Name',
    'Last Name',
    'Address',
    'City',
    'Country',
    'Phone',
    'Email',
  ];
  const formState = [
    formFirstName,
    formLastName,
    formAddress,
    formCity,
    formCountry,
    formPhone,
    formEmail,
  ];

  const onFormChange = (event) => {
    const str = event.target.value;
    const field = event.target.name;

    switch (field) {
      case 'First Name':
        setFormFirstName(str);
        return;
      case 'Last Name':
        setFormLastName(str);
        return;
      case 'Address':
        setFormAddress(str);
        return;
      case 'City':
        setFormCity(str);
        return;
      case 'Country':
        setFormCountry(str);
        return;
      case 'Phone':
        setFormPhone(str);
        return;
      case 'Email':
        setFormEmail(str);
        return;
      default:
        return;
    }
  };

  const onFormSubmit = useCallback(
    (event) => {
      event.preventDefault();

      // dispatch to save contact info
      // pass in formState array
    },
    [...formState]
  );

  return (
    <section className="contact-details">
      <form>
        {formFields.map((field, index) => {
          <label key={`${index}${field}`}>
            <input
              onChange={(event) => {
                onFormChange(event);
              }}
              placeholder={`${field}`}
              required
              type="text"
              value={formState[index]}
            />
            ;
          </label>;
        })}
      </form>
      <button className="cancel-button">Cancel</button>
      <button className="save-button">Save</button>
    </section>
  );
};

export default ContactDetails;
