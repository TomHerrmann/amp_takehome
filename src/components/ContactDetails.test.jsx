import React from 'react';
import { useDispatch, useStore } from 'react-redux';
import { act } from 'react-dom/test-utils';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import ContactDetails from './ContactDetails.jsx';
import { contactUpdateCancel, contactUpdateSave } from '../actions/actions';
import store from '../__testHelpers__/mockStore';

jest.mock('react-redux');

describe('<ContactDetails />', () => {
  let dispatch;
  beforeEach(() => {
    store.getState().currentContact = store.getState().contacts[0];
    useStore.mockImplementation(jest.fn(() => store));

    dispatch = jest.fn();
    useDispatch.mockReset();
    useDispatch.mockReturnValue(dispatch);
  });

  it('should render correctly', () => {
    const component = mount(<ContactDetails />);
    expect(toJson(component)).toMatchSnapshot();
  });
  it('should correctly update the form on an onFormChange event', () => {
    const component = mount(<ContactDetails />);
    act(() => {
      component
        .find('input')
        .at(0)
        .props()
        .onChange({ target: { name: 'First Name', value: 'Tom' } });
    });
    component.update();
    expect(
      component
        .find('input')
        .at(0)
        .props().value
    ).toEqual('Tom');
  });
  it('should dispatch the correct action on onFormCancel event', () => {
    const component = mount(<ContactDetails />);
    component
      .find('[className="cancel-button"]')
      .props()
      .onClick();
    expect(dispatch).toHaveBeenCalledWith(contactUpdateCancel());
  });
  it('should dispatch the correct action on onFormSave event', () => {
    const component = mount(<ContactDetails />);
    component
      .find('[className="save-button"]')
      .props()
      .onClick();
    expect(dispatch).toHaveBeenCalledWith(
      contactUpdateSave({
        address: '42991 Beahan Parkway',
        city: 'Veumberg',
        country: 'Venezuela',
        email: 'Marianne56@hotmail.com',
        firstName: 'Brandy',
        lastName: 'Bartell',
        phone: '586-723-4824',
      })
    );
  });
});
