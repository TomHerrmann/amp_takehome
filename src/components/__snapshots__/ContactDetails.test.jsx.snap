// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContactDetails /> should render correctly 1`] = `
<ContactDetails>
  <section
    className="contact-details"
  >
    <form>
      <label
        className="label-narrow"
        key="0First Name"
      >
        First Name
        <br />
        <input
          name="First Name"
          onChange={[Function]}
          placeholder="Brandy"
          required={true}
          type="text"
          value="Brandy"
        />
      </label>
      <label
        className="label-narrow"
        key="1Last Name"
      >
        Last Name
        <br />
        <input
          name="Last Name"
          onChange={[Function]}
          placeholder="Bartell"
          required={true}
          type="text"
          value="Bartell"
        />
      </label>
      <label
        className="label-wide"
        key="2Address"
      >
        Address
        <br />
        <input
          name="Address"
          onChange={[Function]}
          placeholder="42991 Beahan Parkway"
          required={true}
          type="text"
          value="42991 Beahan Parkway"
        />
      </label>
      <label
        className="label-narrow"
        key="3City"
      >
        City
        <br />
        <input
          name="City"
          onChange={[Function]}
          placeholder="Veumberg"
          required={true}
          type="text"
          value="Veumberg"
        />
      </label>
      <label
        className="label-narrow"
        key="4Country"
      >
        Country
        <br />
        <input
          name="Country"
          onChange={[Function]}
          placeholder="Venezuela"
          required={true}
          type="text"
          value="Venezuela"
        />
      </label>
      <label
        className="label-narrow"
        key="5Phone"
      >
        Phone
        <br />
        <input
          name="Phone"
          onChange={[Function]}
          placeholder="586-723-4824"
          required={true}
          type="text"
          value="586-723-4824"
        />
      </label>
      <label
        className="label-wide"
        key="6Email"
      >
        Email
        <br />
        <input
          name="Email"
          onChange={[Function]}
          placeholder="Marianne56@hotmail.com"
          required={true}
          type="text"
          value="Marianne56@hotmail.com"
        />
      </label>
    </form>
    <section
      className="contact-form-buttons"
    >
      <button
        className="cancel-button"
        onClick={[Function]}
      >
        Cancel
      </button>
      <button
        className="save-button"
        onClick={[Function]}
      >
        Save
      </button>
    </section>
  </section>
</ContactDetails>
`;
